web/src/upload_widget.ts

```diff
@@ -34,70 +245,192 @@ export function build_widget(
+                                } else {
+                                    void accept(edited_file);
+                                }
+                                reset_scale = !reset_scale;
+                            } else {
+                                edited_file = new File([file.data], file.name);
+                                const image_blob = URL.createObjectURL(edited_file);
+                                $preview_image.attr("src", image_blob);
+                                uppy.close();
+                                $input_error.hide();
+                                $save_button.hide();
+                                $scale_to_fit_button.hide();
+                                $reset_scale_to_fit_button.hide();
+                                $preview_image?.show();
+                                $file_name_field.show();
+                                $other_elements_to_hide?.show();
+                                if (!$file_name_field.closest("#emoji-file-name").length) {
+                                    $clear_button.hide();
+                                }
+                            }
+                        }
+                    });
+
+                uppy.addFile({
+                    source: "file input",
+                    name: file.name,
+                    type: file.type,
+                    data: file,
+                });
+
+                display_dashboard_after_timeout(true);
+            }
         }
     }
 
     function clear(): void {
+        if (uppy) {
+            uppy.close();
+        }
         const $control = get_file_input();
         $control.val("");
         $file_name_field.text("");
+        $save_button.hide();
+        $scale_to_fit_button.hide();
+        $reset_scale_to_fit_button.hide();
+        $input_error.hide();
+        $other_elements_to_hide?.show();
         $clear_button.hide();
         $upload_button.show();
-        if ($preview_text !== undefined) {
-            $preview_text.hide();
-        }
     }
 
     $clear_button.on("click", (e) => {
         clear();
         e.preventDefault();
     });
 
+    $save_button.on("click", (e) => {
+        $(".uppy-DashboardContent-save").trigger("click");
+        e.preventDefault();
+    });
+
+    $scale_to_fit_button.on("click", (e) => {
+        edited = true;
+        $(".uppy-DashboardContent-save").trigger("click");
+        e.preventDefault();
+    });
+
+    $reset_scale_to_fit_button.on("click", (e) => {
+        edited = true;
+        $(".uppy-DashboardContent-save").trigger("click");
+        e.preventDefault();
+    });
+
     $upload_button.on("drop", (e) => {
         const files = e.originalEvent?.dataTransfer?.files;
         if (files === null || files === undefined || files.length === 0) {
             return false;
         }
-        get_file_input()[0]!.files = files;
+        const $file_input = get_file_input();
```

Comment:
I don't understand why this is updated from e.target.files?.[0] === undefined to this particular thing.